### PR TITLE
sql: remove experimental multi-column inverted index session setting

### PR DESCRIPTION
**Pull request by mgartner**
#### sql: remove experimental multi-column inverted index session setting

The session setting gating access to multi-column inverted indexes is
removed now that the feature is supported. Multi-column inverted indexes
add an additional indexing option for users. They will provide the
foundation to allow for partitioned inverted indexes.

Informs #43643

Release note (sql change): Multi-column inverted indexes can now be
created. The last indexed column must be inverted types such as JSON,
ARRAY, GEOMETRY, and GEOGRAPHY. All preceding columns must have types
that are indexable. These indexes may be used for queries that constrain
all index columns.

#### sql: better error messages for invalid inverted index columns

Release note: None

### CODE REVIEW

**Review by RaduBerinde — COMMENTED** [review]
:lgtm: 

[nit] in the release note, I'd say "all preceding columns must have types that are indexable" and "that constrain all index columns"

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/59565#-:-MSEieSd0LOcQ3d4y7r8:b-z8avy7)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @rytaft)

**Comment by mgartner** [comment]
Done.

File: pkg/sql/catalog/tabledesc/structured.go [hunk 1]
```diff
@@ -2608,27 +2605,41 @@ func checkColumnsValidForIndex(tableDesc *Mutable, indexColNames []string) error
...
-					invalidColumns = append(invalidColumns, col)
+				// The last column indexed by an inverted index must be
+				// inverted indexable.
+				if i == lastCol && !colinfo.ColumnTypeIsInvertedIndexable(col.Type) {
+					return errors.WithHint(
+						pgerror.New(
+							pgcode.FeatureNotSupported,
+							fmt.Sprintf("column %s of type %s is not allowed as the last column in an inverted index", col.Name, col.Type.Name()),
+						),
+						"see the documentation for more information about inverted indexes",
```
Comment (ianjevans) [inline pkg/sql/catalog/tabledesc/structured.go:2626]: "The last column of an inverted index definition must be one of the inverted indexable types: JSON, ARRAY, GEOMETRY, .... See the documentation on INVERTED INDEX for more details."

File: pkg/sql/catalog/tabledesc/structured.go [hunk 2]
```diff
@@ -2608,27 +2605,41 @@ func checkColumnsValidForIndex(tableDesc *Mutable, indexColNames []string) error
...
+
+				}
+				// Any preceding columns must not be inverted indexable.
+				if i < lastCol && !colinfo.ColumnTypeIsIndexable(col.Type) {
+					return errors.WithHint(
+						pgerror.New(
+							pgcode.FeatureNotSupported,
+							fmt.Sprintf("column %s of type %s is only allowed as the last column in an inverted index", col.Name, col.Type.Name()),
+						),
+						"see the documentation for more information about inverted indexes",
```
Comment (ianjevans) [inline pkg/sql/catalog/tabledesc/structured.go:2637]: "Only the last column of an inverted index definition can be an inverted indexable type. Change the order of the inverted index definition and only use one column of inverted indexable type per inverted index. See the documentation on INVERTED INDEX for more details."

**Review by rytaft — APPROVED** [review]
:lgtm: 

> Reviewed 17 of 17 files at r1, 2 of 3 files at r2, 1 of 1 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/59565#-:-MSFdw_x5mSTvni50VO4:bg7igkv)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @ianjevans)

___
*[pkg/sql/catalog/tabledesc/structured.go, line 2626 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/59565#-MSEmtE7nAMnN204oGnC-r2-2626:-MSFcbPbE-d7WvtFc9Vl:bb7lejh) ([raw file](https://github.com/cockroachdb/cockroach/blob/c771667b380ee93151e2217548930b0d8041cb8a/pkg/sql/catalog/tabledesc/structured.go#L2626)):*
<details><summary><i>Previously, ianjevans (Ian Evans) wrote…</i></summary><blockquote>

"The last column of an inverted index definition must be one of the inverted indexable types: JSON, ARRAY, GEOMETRY, .... See the documentation on INVERTED INDEX for more details."
</blockquote></details>

Can you also include a link to the docs? Or if that page isn't available yet, just leave a TODO here?
___
*[pkg/sql/catalog/tabledesc/structured.go, line 2637 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/59565#-MSEmtE7nAMnN204oGnC-r2-2637:-MSFctJk2H2Fjdy7k_QT:b-evnfbn) ([raw file](https://github.com/cockroachdb/cockroach/blob/c771667b380ee93151e2217548930b0d8041cb8a/pkg/sql/catalog/tabledesc/structured.go#L2637)):*
<details><summary><i>Previously, ianjevans (Ian Evans) wrote…</i></summary><blockquote>

"Only the last column of an inverted index definition can be an inverted indexable type. Change the order of the inverted index definition and only use one column of inverted indexable type per inverted index. See the documentation on INVERTED INDEX for more details."
</blockquote></details>

[nit] this suggestion seems a bit long for an error message. Maybe the second sentence isn't necessary since we're already referring users to the docs page?

**Review by mgartner — COMMENTED** [review]
I've made some changes based on the feedback. To be clear, this is what the error messages will look like from the cockroach SQL shell:

```
demo@127.0.0.1:26257/defaultdb> CREATE TABLE t (a INT, j JSON, INVERTED INDEX (j, a));
ERROR: column j of type jsonb is only allowed as the last column in an inverted index
SQLSTATE: 0A000
HINT: Only the last column of an inverted index can be of type jsonb.
See the documentation on INVERTED INDEX for more details: https://www.cockroachlabs.com/docs/stable/inverted-indexes.html

demo@127.0.0.1:26257/defaultdb> CREATE TABLE t (a INT, j JSON, INVERTED INDEX (a));
ERROR: column a of type int is not allowed as the last column in an inverted index
SQLSTATE: 0A000
HINT: The last column of an inverted index definition must be one of the inverted indexable types: JSONB, ARRAY, GEOMETRY, or GEOGRAPHY.
See the documentation on INVERTED INDEX for more details: https://www.cockroachlabs.com/docs/stable/inverted-indexes.html
```

@ianjevans @rytaft What do you think?

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/59565#-:-MSTnIFM7hsNTnSZIjHg:b8dal2)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @ianjevans)

**Comment by ianjevans** [comment]
LGTM, Marcus.

**Review by rytaft — APPROVED** [review]
:lgtm: 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/59565#-:-MSTzPKw0xjnk5WmviGH:bg7igkv)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @ianjevans)

**Comment by mgartner** [comment]
TFTRs!

bors r+

**Pull request merged**